selfservice/strategy/oidc/provider_lark.go

```diff
@@ -0,0 +1,177 @@
+	config *Configuration
+	reg    dependencies
+}
+
+func NewProviderLark(
+	config *Configuration,
+	reg dependencies,
+) *ProviderLark {
+	return &ProviderLark{
+		config: config,
+		reg:    reg,
+	}
+}
+
+func (g *ProviderLark) Config() *Configuration {
+	return g.config
+}
+
+func (g *ProviderLark) OAuth2(ctx context.Context) (*oauth2.Config, error) {
+	var endpoint = oauth2.Endpoint{
+		AuthURL:   "https://passport.feishu.cn/suite/passport/oauth/authorize",
+		TokenURL:  "https://passport.feishu.cn/suite/passport/oauth/token",
+		AuthStyle: oauth2.AuthStyleInParams,
+	}
+
+	return &oauth2.Config{
+		ClientID:     g.config.ClientID,
+		ClientSecret: g.config.ClientSecret,
+		Endpoint:     endpoint,
+		// DingTalk only allow to set scopes: openid or openid corpid
+		Scopes:      g.config.Scope,
+		RedirectURL: g.config.Redir(g.reg.Config().OIDCRedirectURIBase(ctx)),
+	}, nil
+
+}
+
+func (g *ProviderLark) Claims(ctx context.Context, exchange *oauth2.Token, query url.Values) (*Claims, error) {
+	// larkClaim is defined in the https://open.feishu.cn/document/common-capabilities/sso/api/get-user-info
+	type larkClaim struct {
+		Sub          string `json:"sub"`
+		Name         string `json:"name"`
+		Picture      string `json:"picture"`
+		OpenID       string `json:"open_id"`
+		UnionID      string `json:"union_id"`
+		EnName       string `json:"en_name"`
+		TenantKey    string `json:"tenant_key"`
+		AvatarURL    string `json:"avatar_url"`
+		AvatarThumb  string `json:"avatar_thumb"`
+		AvatarMiddle string `json:"avatar_middle"`
+		AvatarBig    string `json:"avatar_big"`
+		Email        string `json:"email"`
+		UserID       string `json:"user_id"`
+		Mobile       string `json:"mobile"`
+	}
+	var (
+		userEndpoint = "https://passport.feishu.cn/suite/passport/oauth/userinfo"
+		accessToken  = exchange.AccessToken
+		client       = g.reg.HTTPClient(ctx, httpx.ResilientClientDisallowInternalIPs())
+		user         larkClaim
+	)
+
+	req, err := retryablehttp.NewRequest("GET", userEndpoint, nil)
+	if err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+
+	req.Header.Add("Authorization", "Bearer "+accessToken)
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+	defer resp.Body.Close()
+
+	if err := json.NewDecoder(resp.Body).Decode(&user); err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+
+	return &Claims{
+		Issuer:      userEndpoint,
+		Subject:     user.OpenID,
+		Name:        user.Name,
+		Nickname:    user.Name,
+		Picture:     user.AvatarURL,
+		Email:       user.Email,
+		PhoneNumber: user.Mobile,
+	}, nil
+
+}
+
+func (pl *ProviderLark) AuthCodeURLOptions(r ider) []oauth2.AuthCodeOption {
+	return []oauth2.AuthCodeOption{}
+}
+
+func (g *ProviderLark) Exchange(ctx context.Context, code string, opts ...oauth2.AuthCodeOption) (*oauth2.Token, error) {
+
```

Comment:
Why can't you use the normal `g.oauth2()` flow here? Is there a reason why we need to override this? I don't see any big differences.

https://github.com/ory/kratos/blob/b3b6f88d61652d8697288824a9dda3dd55148ccc/selfservice/strategy/oidc/provider_gitlab.go#L67

Reply:
Sorry, I am not familiar with this package.

You are right. Using the exchanger in the `oauth2` package is a good idea.

I have updated the code.